### ISSUE TITLE
DataSourceFactory with OkHttpDataSource ignored

### ISSUE THREAD

**Issue by Nucifera8472** [androidx/media#116]
### Media3 Version

1.0.0-beta01

### Devices that reproduce the issue

- Samsung Galaxy A52s
- bq Aquaris X Pro

### Reproducible in the demo app?

Not tested

### Reproduction steps

I am trying to switch the network stack to OkHttp as described in the Exoplayer documentation. The player instance is created inside a MediaSessionService with a custom mediaSourceFactory containing the OkHttpDataSouce like so:

```
        player = ExoPlayer.Builder(baseContext)
            .setMediaSourceFactory(
                DefaultMediaSourceFactory(baseContext)
                    .setDataSourceFactory {
                        // this code is never called -----------------------------------------------
                        OkHttpDataSource.Factory(okHttpClient).createDataSource() 
                    }
            )
            .build()
```

The mediaSource is never directly set anywhere else in the code, the only interaction is via the mediaController which I use to set MediaItem playlists.

```
          val requestMetadata = MediaItem.RequestMetadata.Builder()
              .setMediaUri(item.url.toUri())
              .build()
          val mediaItem = MediaItem.Builder()
              .setMediaId(item.id)
              .setRequestMetadata(requestMetadata)
              .build()
        val mediaItems = listOf(mediaItem)  // playlist generation code omitted for brevity
        mediaController?.setMediaItems(mediaItems, startIndex, C.TIME_UNSET)
```

### Expected result

Player is expected to use the provided OkHttpClient for network connection.

### Actual result

According to the logs, the player is still using DefaultHttpDataSource for communication instead of the provided OkHttpDataSource. The line of code where the actual data source creation by the factory would happen never gets called. 

The reason I am trying to switch the network stack is that some of the online radio URLs that I try to play return 301 or 302 HTTP status codes. While information I found suggests that redirects should be working, I thought that maybe the DefaultHttpDataSource is one that doesn't handle it well at that OkHttp would handle it better. And if not I at least wanted to find out what is going on in the redirect which I thought would be easy by using an OkHttpClient with a logging interceptor. I also tried enabling cross protocol redirects (`setAllowCrossProtocolRedirects(true)` ) on the DefaultHttpDataSource, but it didn't work - which I assume is also caused by the datasource being ignored.

```
DefaultMediaSourceFactory(baseContext)
    .setDataSourceFactory(
        DefaultHttpDataSource.Factory().setAllowCrossProtocolRedirects(true)
    )
```

### Media

The issue isn't caused by any specific media type, but that's the one I was testing it with:

Original URL http://stream.mangoradio.de/ (ERROR_CODE_IO_BAD_HTTP_STATUS, HTTP 301) 
Redirected URL https://stream.laut.fm/mangoradio (this one plays just fine)

### Bug Report

- [ ] You will email the zip file produced by `adb bugreport` to dev.exoplayer@gmail.com after filing this issue.

**Comment by tonihei** [comment]
Thanks for reporting! We discovered the same bug ourselves yesterday and it will be fixed soon.

**Comment by tonihei** [comment]
Fix by the commit above, will be released soon in ExoPlayer 2.18.1 / Media3 1.0.0-beta02.
